### ISSUE TITLE
http: panic serving 127.0.0.1:53816

### ISSUE THREAD

**Issue by ionics** [DRuggeri/alertmanager_gotify_bridge#9]
Hi folks,

please give advice, what could be the cause of the following panic.

### Commands:
Command to start alertmanager_gotify_bridge:

1) 

`export GOTIFY_TOKEN=Az2xthObihA****
`

2) 

```
alertmanager_gotify_bridge --gotify_endpoint="http://127.0.0.1:9099/message" --bind_address=0.0.0.0 --port=8080 --webhook_path="/gotify_webhook" --timeout=2s --title_annotation="description" --message_annotation="summary" --priority_annotation="priority" --default_priority=5 --metrics_namespace="alertmanager_gotify_bridge" --metrics_path="/metrics" --extended_details --debug
Starting debug server on http://0.0.0.0:8080/gotify_webhook translating to http://127.0.0.1:9099/message
```

### Panic occurs:

As you can see a test alert created with test-alert.script.sh is successfully pushed from alertmanager to alertmanager_gotify_bridge but a panic occurs.
```
2021/08/03 11:58:14 Detected 1 alerts
2021/08/03 11:58:14   Alert 0
2021/08/03 11:58:14     title annotation (description) missing
2021/08/03 11:58:14     message: High latency is high!
2021/08/03 11:58:14     priority annotation (priority) missing - falling back to default (5)
2021/08/03 11:58:14 http: panic serving 127.0.0.1:53816: assignment to entry in nil map
goroutine 21 [running]:
net/http.(*conn).serve.func1(0xc000159180)
	/usr/local/go/src/net/http/server.go:1800 +0x139
panic(0x8f1820, 0xa2b950)
	/usr/local/go/src/runtime/panic.go:975 +0x3e3
main.(*bridge).handleCall(0xc0000b8b90, 0xa3d3e0, 0xc0000b61c0, 0xc00015c500)
	/home/hax/go/src/github.com/DRuggeri/alertmanager_gotify_bridge/main.go:307 +0x1a29
net/http.HandlerFunc.ServeHTTP(0xc000099bb0, 0xa3d3e0, 0xc0000b61c0, 0xc00015c500)
	/usr/local/go/src/net/http/server.go:2041 +0x44
net/http.(*ServeMux).ServeHTTP(0xc0000cd200, 0xa3d3e0, 0xc0000b61c0, 0xc00015c500)
	/usr/local/go/src/net/http/server.go:2416 +0x1a5
net/http.serverHandler.ServeHTTP(0xc0000b60e0, 0xa3d3e0, 0xc0000b61c0, 0xc00015c500)
	/usr/local/go/src/net/http/server.go:2836 +0xa3
net/http.(*conn).serve(0xc000159180, 0xa3e8e0, 0xc0000cd280)
	/usr/local/go/src/net/http/server.go:1924 +0x86c
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2962 +0x35c
```

### For your information the web.hook part in alertmanager.yml file:
```
TLDR;....
- name: 'web.hook'
  webhook_configs:
  #- url: 'http://127.0.0.1:5001/'
  - url: http://127.0.0.1:8080/gotify_webhook
    send_resolved: true
TLDR;....
```

### gotify is running successfully and listening on Port 9099

`
tcp        0      0 127.0.0.1:9099          0.0.0.0:*               LISTEN      21390/gotify 
`

In gotify webui i have created an App named alertmanager and  the token provided is used as 'export GOTIFY_TOKEN=*****'

Any hints what i could have done wrong?

The issue seems related to startup command flag "--extended_details" due to line 307 in main.go

```
			if *extendedDetails {
				if strings.HasPrefix(alert.GeneratorURL, "http") {
					message += "<br/><a href='" + alert.GeneratorURL + "'>go to source</a>"
					extrasNotification := make(map[string]map[string]string)
	307 ===> 		extrasNotification["click"]["url"] = alert.GeneratorURL
					extras["client::notification"] = extrasNotification
				}
```

but if i do remove --extended_details the error changes to "Unable to dispatch!"

```
2021/08/03 12:24:08 bridge: data sent - unmarshalling from JSON: {"receiver":"web\\.hook","status":"firing","alerts":[{"status":"firing","labels":{"alertname":"fooAlert-9143","instance":"fooAlert-9143.example.net","label_costcentre":"FOO","namespace":"foo-bar","service":"my-service","severity":"warning"},"annotations":{"summary":"High latency is high!"},"startsAt":"2021-08-03T12:23:38Z","endsAt":"0001-01-01T00:00:00Z","generatorURL":"http://localhost/fooAlert-9143","fingerprint":"dab0433fc03b7200"}],"groupLabels":{"alertname":"fooAlert-9143"},"commonLabels":{"alertname":"fooAlert-9143","instance":"fooAlert-9143.example.net","label_costcentre":"FOO","namespace":"foo-bar","service":"my-service","severity":"warning"},"commonAnnotations":{"summary":"High latency is high!"},"externalURL":"http://0.0.0.0:9093","version":"4","groupKey":"{}:{alertname=\"fooAlert-9143\"}","truncatedAlerts":0}

2021/08/03 12:24:08 Detected 1 alerts
2021/08/03 12:24:08   Alert 0
2021/08/03 12:24:08     title annotation (description) missing
2021/08/03 12:24:08     message: High latency is high!
2021/08/03 12:24:08     priority annotation (priority) missing - falling back to default (5)
2021/08/03 12:24:08     Unable to dispatch!

```

Thanks in advance,
kind regards,
Raphael

**Comment by ionics** [comment]
SOLUTION:

my used start command especially:

`--title_annotation="title" --message_annotation="description" --priority_annotation="priority"`

did not match up correctly.

### Explanation:

gotify API is expecting --title_annotation, --message_annotation and --priority in my alert_rules.yml located unter /etc/prometheus/rules/alert_rules_yourdomain.yml and example rule is defined as follows:

```
- name: Hardware alerts
  rules:
  - alert: Node down
    expr: up{job="node"} == 0
    for: 1m
    labels:
      severity: high
    annotations:
      priority: 5
      summary: Node {{ $labels.instance }} is down
      description: Failed to scrape {{ $labels.job }} on {{ $labels.instance }} for more than 3 minutes. Node seems down.     
```

so to match up with gotify API:

- --title_annotation="summary"
- --message_annotation="description"
- --priority_annotation="priority"

such explanations need to match.

Do not expect gotify API to accept bad formatted messages. You would end up with alertmanager_gotify_bridge to error out with 'Unable to dispatch'... a better warning message such as ... "gotify API bad formated message would be hintful."

Closed. 

Kind regards,
Raphael
